### PR TITLE
Added support for culture neutral patterns

### PR DESCRIPTION
**Pull request by MarcoViglione-Laser**
Fixed support for localized and culture neutral patterns (issue #7430)

### CODE REVIEW

File: src/Orchard.Web/Modules/Orchard.Autoroute/Services/AutorouteService.cs [hunk 1]
```diff
@@ -135,32 +135,44 @@ public class AutorouteService : Component, IAutorouteService {
...
+                    }
+                    else {
+                        return new RoutePattern { Name = "Title", Description = "my-title", Pattern = "{Content.Slug}", Culture = culture };
+                    }
                 }
-            }
 
-            // Return a default pattern if set.
-            var patternCultureSearch = settings.Patterns.Any(x => String.Equals(x.Culture, culture, StringComparison.OrdinalIgnoreCase)) ? culture : null;
+                // If patterns for the specified culture exist I search one of them, else I search a culture neutral pattern.
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Services/AutorouteService.cs]: Make your comments impersonal: no "I", "He", "She" ...

File: src/Orchard.Web/Modules/Orchard.Autoroute/Migrations.cs [hunk 2]
```diff
@@ -49,5 +62,46 @@ public class Migrations : DataMigrationImpl {
 
             return 4;
         }
+
+        public int UpdateFrom4() {
+            // Adding some culture neutral patterns if they don't exist
+            var types = _contentManager.GetContentTypeDefinitions().Where(t => t.Parts.Any(p => p.PartDefinition.Name.Equals(typeof(AutoroutePart).Name)));
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Migrations.cs]: You can do the test in the loop, as you are also looking for the part later.

File: src/Orchard.Web/Modules/Orchard.Autoroute/Migrations.cs [hunk 3]
```diff
@@ -49,5 +62,46 @@ public class Migrations : DataMigrationImpl {
...
+            foreach (var type in types) {
+                var typeDefinition = ContentDefinitionManager.GetTypeDefinition((type.Name));
+                if (typeDefinition != null) {
+                    var settings = typeDefinition.Parts.First(x => x.PartDefinition.Name == "AutoroutePart").Settings.GetModel<AutorouteSettings>();
+                    if (!settings.Patterns.Any(x => String.IsNullOrWhiteSpace(x.Culture))) {
+                        string siteCulture = _cultureManager.GetSiteCulture();
+                        string newPatterns = "";
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Migrations.cs]: Make `newPatterns` a `List<string>` then at the end you can just join all of them using `String.Join(", ")`. This way you don't have to deal with concatenating `,` and then having to remove the last one.

File: src/Orchard.Web/Modules/Orchard.Autoroute/Migrations.cs (hunk 2 above)
Comment (HermesSbicego-Laser) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Migrations.cs]: It seems to us that filtering types having AutoroutePart before looping is more efficient, or are we misunderstanding your comment?

Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Migrations.cs]: var typeDefinitions = ContentDefinitionManager.ListTypeDefinition( ... where has autoroute part);

Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Migrations.cs]: I am sure you can even get the ContentTypePartDefinition with AutoroutePart for all type

**Comment by Skrypt** [comment]
looks like good work will test it

**Comment by Skrypt** [comment]
I don't know why but I can't see the culture neutral tab. You guys took another direction?

**Comment by MarcoViglione-Laser** [comment]
The culture neutral tab is called "All cultures", you can find it in the content definition.

cultureneutraltab

**Comment by Skrypt** [comment]
Sorry I presumed it was on the dev branch but it's not been merged yet from 1.10.x branch. Tested it on 1.10.x branch and I see it now.

**Pull request merged**